habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -469,6 +469,178 @@ def step(self, *args, is_last_action, **kwargs):
+        new_target_trans = mn.Matrix4.from_(
+            target_rigid_state.rotation.to_matrix(), new_end_pos
+        )
+
+        # For detection of linear or angualr velocities
+        # There is a collision if the difference between the clamped NavMesh position and target position is too great for any point.
+        lin_diff = len([v for v in move if v > self._lin_collision_threshold])
+        ang_diff = len([v for v in move if v > self._ang_collision_threshold])
+
+        if (lin_diff > 0) and center_move != 0:
+            return True, new_target_trans
+        elif (ang_diff > 0) and center_move == 0:
+            return True, new_target_trans
+        else:
+            return False, new_target_trans
+
+    def update_base(self):
+        ctrl_freq = self._sim.ctrl_freq
+
+        trans = self.cur_articulated_agent.sim_obj.transformation
+        rigid_state = habitat_sim.RigidState(
+            mn.Quaternion.from_matrix(trans.rotation()), trans.translation
+        )
+
+        target_rigid_state = self.base_vel_ctrl.integrate_transform(
+            1 / ctrl_freq, rigid_state
+        )
+        end_pos = self._sim.step_filter(
+            rigid_state.translation, target_rigid_state.translation
+        )
+
+        end_pos -= self.cur_articulated_agent.params.base_offset
+        target_trans = mn.Matrix4.from_(
+            target_rigid_state.rotation.to_matrix(), end_pos
+        )
+
+        did_coll, new_target_trans = self.collision_check(
+            trans, target_trans, end_pos, target_rigid_state
+        )
+        self.cur_articulated_agent.sim_obj.transformation = new_target_trans
+
+        if not self._allow_dyn_slide:
+            if did_coll:
+                self.cur_articulated_agent.sim_obj.transformation = trans
+        if self.cur_grasp_mgr.snap_idx is not None:
+            # Holding onto an object, also kinematically update the object.
+            # object.
+            self.cur_grasp_mgr.update_object_to_grasp()
+
+        if self.cur_articulated_agent._base_type == "leg":
+            # Fix the leg joints
+            self.cur_articulated_agent.leg_joint_pos = (
+                self.cur_articulated_agent.params.leg_init_params
+            )
+
+    def step(self, *args, is_last_action, **kwargs):
+        lateral_lin_vel = 0.0
+        if self._enable_lateral_move:
+            lin_vel, lateral_lin_vel, ang_vel = kwargs[
+                self._action_arg_prefix + "base_vel"
+            ]
+        else:
+            lin_vel, ang_vel = kwargs[self._action_arg_prefix + "base_vel"]
+
+        lin_vel = np.clip(lin_vel, -1, 1) * self._lin_speed
+        lateral_lin_vel = np.clip(lateral_lin_vel, -1, 1) * self._lin_speed
+        ang_vel = np.clip(ang_vel, -1, 1) * self._ang_speed
+        if not self._allow_back:
+            lin_vel = np.maximum(lin_vel, 0)
+
+        self.base_vel_ctrl.linear_velocity = mn.Vector3(
+            lin_vel, 0, -lateral_lin_vel
```

Comment:
why `-lateral_lin_vel`?

Reply:
This is due to the coordinate of Habitat, so we add negative sign here. Thank you!